src/fontlibc/fontlibc.h

```diff
@@ -0,0 +1,437 @@
+
+typedef enum {
+	/* clear = sans-serif font */
+	SERIF = 0x01,
+	/* If both are set, then assume there's no difference between oblique
+	 * and italic styles. */
+	OBLIQUE = 0x02,
+	ITALIC = 0x04,
+	/* Chances are you're not using this library for monospaced fonts.
+	 * But if you are, you'll still have to provide a widths table where
+	 * every byte is the same. */
+	MONOSPACED = 0x08
+} font_styles_t;
+
+typedef struct {
+	/* These are standard C-strings.  These may be NULL. */
+	intptr_t font_family_name;
+	intptr_t font_author;
+	/* NOTA BENE: TYPEFACES AND BITMAPPED FONTS CANNOT BE COPYRIGHTED UNDER U.S. LAW!
+	 * This field is therefore referred to as a pseudocopyright.  HOWEVER,
+	 * it IS is applicable in other jusrisdictions, such as Germany. */
+	intptr_t font_pseudocopyright;
+	intptr_t font_description;
+	intptr_t font_version;
+	intptr_t font_codepage;
+} font_metadata;
+
+typedef struct {
+	/* Must be zero */
+	uint8_t fontVersion;
+	/* Height in pixels */
+	uint8_t height;
+	/* Total number of glyphs provided. */
+	uint8_t total_glyphs;
+	/* Number of first glyph.  If you have no codepoints below 32, for
+	   example, you can omit the first 32 bitmaps. */
+	uint8_t first_glyph;
+	/* Offset/pointer to glyph widths table.
+	 * This is an OFFSET from the fontVersion member in data format.
+	 * However, it is 24-bits long because it becomes a real pointer upon loading. */
+	intptr_t widths_table;
+	/* Offset to a table of offsets to glyph bitmaps.
+	 * These offsets are only 16-bits each to save some space. */
+	intptr_t bitmaps;
+	/* Specifies how much to move the cursor left after each glyph.
+	   Total movement is width - overhang.  Intended for italics. */
+	uint8_t italic_space_adjust;
+	/* These suggest adding blank space above or below each line of text.
+	   This can increase legibility. */
+	uint8_t space_above;
+	uint8_t space_below;
+	/* Specifies the boldness of the font. 
+	   0x40: light
+	   0x80: regular
+	   0x90: medium
+	   0xC0: bold*/
+	uint8_t weight;
+	/* Specifies the style of the font.  See enum font_styles */
+	uint8_t style;
+	/* For layout, allows aligning text of differing fonts vertically.
+	   These count pixels going down, i.e. 0 means the top of the glyph. */
+	uint8_t cap_height;
+	uint8_t x_height;
+	uint8_t baseline_height;
+} raw_font;
+
+typedef struct {
+	char header[8]; /* "FONTPACK" */
+	/* Frankly, if you have more than 127 fonts in a pack, you have a
+	   problem. */
+	uint8_t fontCount;
+	/* Offset from first byte of header */
+	intptr_t metadata;
+	intptr_t font_list;
+} font_pack;
+
+
+/**
+ * Sets the bounds of the window all text will appear in to be the full screen
+ */
+void fontlib_SetWindowFullScreen(void);
+
+/**
+ * Sets the bounds of the window all text will appear in
+ * @param x_min X coord base
+ * @param y_min Y coord base
+ * @param width Width
+ * @param height Height
+ */
+void fontlib_SetWindow(int x_min, uint8_t y_min, int width, uint8_t height);
```

Comment:
I'd suggest adding a `fontlib_window_t` struct that encapsulates these four values and using that instead. If you match the layout in the library, then this functions becomes rather trivial in implementation. `fontlib_GetWindow` would be rather trivial as well, and I'd recommend resurrecting that.

Reply:
And I'm not entirely sure how ZDS's codegen works when constructing temporary structs, but having the parameter be a `const fontlib_window_t *` might be more optimal.

Reply:
Oh, and it's a bit weird having only a bulk setter while having individual getters. Unless there's a good case to be made for the individual getters, I'd suggest removing those (once `fontlib_GetWindow` has been resurrected).

Reply:
_(So many comments...)_ If you're not planning to support clipping, `x_min` and `width` should be of type `uint24_t`. If you are, then `y_min` and `height` should be of type `int`.

Reply:
Internally, it stores an XMax and YMax, not width and height. The former is more useful for bounds checking, the latter seems more useful for actual layout. I think the individual getters are probably faster to call. I didn't add individual setters because it seems unlikely to be frequently useful, but it at least it wouldn't be hard to add.

Reply:
You can still have the public struct use width and height and internally store it as xmax and ymax. The implementation of the bulk getter/setter is no longer a simple copy, but it's still simpler (or at least no more complex) than it is currently. And I'd say that both are simpler from the user's perspective.

Regarding individual getters/setters: neither seems particularly useful to me. If anything the getters seem even less useful than the setters. After all, the programmer set the bounds; they should already know what they are.

Reply:
I thought about a struct, but internally they're not packed. I think there's even some code in there that depends on being able to write zeros to subsequent bytes for the Y values.

Reply:
That's fine. Due to the width/height and xmax/ymax mismatch, you already need a different underlying representation and will have to keep basically the same underlying implementation. But bundling it all into one struct for bulk operations makes for a better API to the programmer, IMO. (Especially for the bulk getter, which you removed because it was so cumbersome)

Reply:
Actually, I removed it just because it looked slower than just doing a bunch of `var = GetValue();`.

I agree, however, that the struct would make for a better API.

Reply:
The reason I didn't make X and width unsigned is simply that I figured people would mostly be using `int`, signed, as it is the default. While having negative coordinates makes no sense, neither does having an X greater than 320, even though the vast majority of the domain for both `int` and `unsigned int` is outside the range [0,320).